### DISCUSSION TITLE
Conversion from RGB to Hexadecimal might generate misleading values.

### DISCUSSION THREAD

**Discussion by thomasmansencal** [colour-science/colour#744] [Ideas]
So this is not really a defect per-see but more a consequence of working with unbounded values while still being able to produce colours for the web. I noticed it in the past and it has bitten me recently while answering that SO question: https://stackoverflow.com/questions/61455618/how-to-convert-cie-color-in-the-following-format-x-0-615-y-0-346-5-6-cd-m2/ 

I think there is a virtue with being able to convert even HDR and WGC in a decent way to Hexadecimal, at the very least, a red HDR colour should not come up as green!

I would be leaning toward something like that:

- RGB in Range [0, 1]:
  - Nothing to be done
- RGB in Range [-inf, _inf]:
  - Warning
  - Naive Clip
- RGB in Range [1, 80]:
  - Warning
  - Normalise luminance by 80 (sRGB Display Peak Luminance)
- RGB in Range [1, +inf]:
  - Warning
  - Normalise luminance by max(RGB)

This is especially relevant with the Automatic Colour Conversion Graph.

**Comment by thomasmansencal** [comment]
I actually went for simpler:

```python
    if np.any(RGB < 0):
        usage_warning(
            '"RGB" array contains negative values, those will be clipped, '
            'unpredictable results may occur!')

        RGB = np.clip(RGB, 0, np.inf)

    if np.any(RGB > 1):
        usage_warning(
            '"RGB" array contains values over 1 and will be normalised, '
            'unpredictable results may occur!')

        RGB = eotf_inverse_sRGB(normalise_maximum(eotf_sRGB(RGB)))

```

**Comment by sobotka** [comment]
Is this seemingly inconsistent behaviour?

Gamut clipping on low and not on high would result in inconsistent differences between input chromaticity and output for either case?

**Comment by thomasmansencal** [comment]
The negative clipping handles (naively) gamut mapping and the positive normalisation handles potential HDR values, they are separate problems. There is no expectation to be able to roundtrip values and HEX input can only be in the range [0-1].

**Comment by sobotka** [comment]
I believe the “correct” output should be against the source gamut; that is, where values are negative *or* greater than source RGB, the destination values are out of gamut. Either outlier indicates out of gamut results, and I believe the proper approach would be to clip according to the resultant value exceeding the initial values?

**Comment by thomasmansencal** [comment]
The correct output is the one that will produce colours close to what they should be with all the caveats and warnings issued.

Keeping in mind that we don't have gamut mapping, this would mean that whenever a user uses those xy chromaticity coordinates/ CIE xyY values: `[0.615, 0.346, 5.6]`

![image](https://user-images.githubusercontent.com/99779/80896824-9d081080-8d46-11ea-932d-8b6bc478fded.png)

He does not get this colour out:

![](https://dummyimage.com/288x288/d9f14f/000000.png&text=Old%20Behaviour)

Not this one either:

![](https://dummyimage.com/288x288/fef14f/000000.png&text=Clipped%20[0,%201])

But that one:

![](https://dummyimage.com/288x288/fe360b/000000.png&text=New%20Behaviour)

We cannot presume that the user Y value in the SO example is representing absolute luminance values, so divided by 80 is actually not an option.

**Comment by sobotka** [comment]
Agree on the absolute. Agree on the “better” gamut clip results here.

I am curious if there is a difference between the gamut clip on zero *before* versus *after* normalization?

Phoof that old behaviour. 🤣

**Comment by thomasmansencal** [comment]
I haven't tried! To be fair, I'm expecting the users to have ensured that their values are representable properly, the new behaviour is mostly a cheap remedy that comes with some benefits:

- It warns users that they might get unexpected results which they did not get at all, got caught pants down myself on SO :)
- It allows the generation of values that, if not correct, are at least in the ballpark.
- It allows the Automatic Colour Conversion Graph to produce ballpark values too.

```
/Users/kelsolaar/Documents/Development/colour-science/colour/colour/utilities/verbose.py:237: ColourUsageWarning: "RGB" array contains values over 1 and will be normalised, unpredictable results may occur!
  warn(*args, **kwargs)
'#fe360b'
```

Don't get me wrong, it is not ideal but it is certainly not worse! :)

**Comment by sobotka** [comment]
Yep. I think it’s more of a thought process at this point. As in “if normalizing, should we normalize against peak Y = 1.0?” Which in turn implies normalizing against R=G=B == Y = 1.0.

Side request: Can this be wired into the star sRGB star functions?

Graph plotting explodes when sRGB is out of gamut.
